src/variables.ts

```diff
@@ -3,26 +3,36 @@ import type { SchemaLike } from './introspection';
 import type { DocumentNodeLike } from './parser';
 import type { obj } from './utils';
 
+type GetInputObjectType<InputField, Introspection extends SchemaLike> = InputField extends {
+  name: any;
+  type: any;
+}
+  ? InputField extends { defaultValue?: undefined | null; type: { kind: 'NON_NULL' } }
+    ? {
+        [Name in InputField['name']]: unwrapTypeRec<InputField['type'], Introspection, true>;
+      }
+    : {
+        [Name in InputField['name']]?: unwrapTypeRec<
+          InputField['type'],
+          Introspection,
+          true
+        > | null;
+      }
+  : {};
+
 type getInputObjectTypeRec<
   InputFields,
   Introspection extends SchemaLike,
+  IsOneOf extends boolean,
```

Comment:
nit: I'd duplicate `getInputObjectTypeRec` here and hoist this check. I'm unsure whether infinitely mutually recursive inputs break if we don't